### ISSUE TITLE
User account locking after given number of consecutive login failures

### ISSUE THREAD

**Issue by rvasin** [ClickHouse/ClickHouse#54450]
**Use case**
Lock users after given number of failed login attempts.

**Describe the solution you'd like**
MySQL has the user account locking feature and we need about the same feature in ClickHouse. It's a part of ClickHouse certification requirements.

[This article](https://www.percona.com/blog/mysql-8-account-locking/) explains how does it work in MySQL.

In MySQL it's possible:
```sql
CREATE USER ‘percona’@’localhost’ IDENTIFIED BY ‘password’ FAILED_LOGIN_ATTEMPTS 4 PASSWORD_LOCK_TIME 3;
```
Where is `PASSWORD_LOCK_TIME` is number of days. Probably in ClickHouse it should be number of seconds (this is the topic to discuss). This way we may lock the user for given number of hours, for example.

Probably in the initial version we may implement it without the usage of `FAILED_LOGIN_ATTEMPTS` and `PASSWORD_LOCK_TIME` parameters in `CREATE USER`, for example, we can simply add the parameters in `<users>` sections or even better in `<profiles>` sections of `users.xml`.

In MySQL the lock status is saved in `account_locked` column of `mysql.user` system table. In ClickHouse `system.users` table is auto-generated. We cannot save information there.

Therefore the main problem is where to save `account_locked` and other needed columns (like number of failed attempts)? 

## Variant 1: new `system.user_status` system table
Right now most of system tables like `query_log` have `SystemLog` as base class and it's possible only to append log data using `flushImpl()`. 

In PR #54410 @sirvickr is implementing `SystemLogStorage` and derived `BackupsStorage` class with:
1. `flushImpl()` to append data
2. `update()` to update data. 
The same approach we may use in new `UserStatus` system table as derived class of `SystemLogStorage` base class. 

## Variant 2: use existing `system.session_log` table
We may run a complex SQL query on `system.session_log` table to calculate number of failed attempts etc. But this is not probably the best way to run the complex query on non-production level `system.session_log` table for checking user lockout status. Second: it's possible to turn off `system.session_log`  in `config.xml`. 

@Demilivor is already working on this feature. The authorizations part of work is clear how to implement. The unclear part of about the storage. 

@tavplubix do you have any suggestions about which storage to use for `account_locked` etc? 

@den-crane may you can remember how it's implemented in other DBMSs? 

cc @alexey-milovidov, @rschu1ze

**Comment by alexey-milovidov** [comment]
Implement it without persistence - the data will be kept in memory and not saved on restart.

**Comment by alexey-milovidov** [comment]
Anything other than trivial is not needed.

**Comment by alexey-milovidov** [comment]
If you'll need persistence, let's implement quotas in Keeper instead.

**Comment by rvasin** [comment]
As written in the [article](https://www.percona.com/blog/mysql-8-account-locking/) in MySQL it's not persistent:
>Global Reset
>    A global reset of all accounts occur for any of the following conditions:
>    -    MySQL server restart.
>    -    Execution of FLUSH PRIVILEGES.

It's not clear do we need persistence as the requirement for the certification or not (we will ask our staff).

Could you give any hints about "quotas in Keeper" how should it work? Which parts of the source code to look? Is it secure to store it in Keeper (like any user cannot change it)?

**Comment by alexey-milovidov** [comment]
By the way, what type of certification is needed?

> Could you give any hints about "quotas in Keeper" how should it work? Which parts of the source code to look? Is it secure to store it in Keeper (like any user cannot change it)?

Don't worry about it. First, let's add a quota on the number of authentication failures, and the problem is solved.

**Comment by rvasin** [comment]
> By the way, what type of certification is needed?
> 

It's Russian ФСТЭК certification.

**Comment by rschu1ze** [comment]
The proposed features invites DoS attacks. Just identify (or guess) someone's user name and run a couple of failed login attempts.

A better approach are exponentially escalating wait times, e.g. wait N^2 sec after the N-th unsuccessful login attempt from the same IP. It's a slight annoyance for legitimate users, and a major obstacle for malicious ones.

**Comment by alexey-milovidov** [comment]
@rschu1ze ClickHouse has quotas keyed by IP so that it will be automatically supported.

**Comment by rvasin** [comment]
With quotes it's easy to implement. But with this way
1. later it will be hard to implement persistence if needed. 
2. It's not clear how to list all blocked users. 
3. Is it easy / or not to unlock user? 

What about the variant:
Have a `map` in global context filled with user statuses (locked/unlocked) and other parameters (like last lock date/time).
In the first PR we may implement it without the persistence. With this approach later it will be possible to add the persistence in one or another way. 
In this case we may do user unlock in MySQL way:
```sql
ALTER USER [IF EXISTS]   
    user_account_name1
ACCOUNT UNLOCK;  

ALTER USER [IF EXISTS]   
    user_account_name1
ACCOUNT LOCK;  

SELECT 
    user, 
    host, 
    account_locked
FROM 
    mysql.user
WHERE 
    account_locked = true
```
In MySQL it's `mysql.user` table. In ClickHouse we could probably to existing auto-generated `system.users` which will lookup the `map` for `account_locked` or maybe new auto-generated (later persistent) `system.user_status` table. 

I am not sure which variant is better. I will invite our architect @aosorgin  to this discussion.

**Comment by alexey-milovidov** [comment]
> later it will be hard to implement persistence if needed.
It's not clear how to list all blocked users.
Is it easy / or not to unlock user?

All are easy and solved out of the box with quotas.
But let's not think about it at all.

> In this case we may do user unlock in MySQL way:

Let's not implement it - this feature (if supported) should be as easy as possible.
(Otherwise - imagine after ten years, we will have to scratch our heads every time remember this feature).

Let's also put it from another perspective - this feature is neither needed nor wanted in ClickHouse, but if we do almost nothing, it is passable...

**Comment by rvasin** [comment]
@alexey-milovidov OK. Thank you very much for explanation. We will implement it with quotas.

**Comment by Demilivor** [comment]
@alexey-milovidov, @tavplubix  Could you please help me implement the solution with quotas?

How can I get `Quota` or/and `EnabledQuota` during authentication? 

When the user is authenticating (see `IAccessStorage::authenticateImpl`) we have no `Session `or `Query` context to get the existing `EnabledQuota` with values.
To create or get `EnabledQuota` from cache, we may use global context, but then we need `QuotaKey`. 
The client doesn't send `QuotaKey` before authentication using a TCP connection.

**Comment by Demilivor** [comment]
I created a draft PR https://github.com/ClickHouse/ClickHouse/pull/54737. Perhaps he can better clarify what the problem with quota_key/client_key is. You can help me with the idea of implementation in this draft PR.

**Comment by rvasin** [comment]
@vitlibar maybe you may help (we see in Git that you implemented many features related to quotas).

**Comment by Demilivor** [comment]
PR #54737 is updated.
I used hardcoded quota key or the authentication to avoid throwing empty key exception in case `QuotaKeyType::CLIENT_KEY`.

**Comment by rvasin** [comment]
@alexey-milovidov We are planning to restart working on the feature. Could you please confirm that implementation using quotas is still the good/actual approach? If it's not persistent - then it's OK for us (for now). 

Also @Demilivor asked for advice about the usage of global context (which is not the best way for quotas). Maybe someone could advise/help to point out the right direction. And we will implement it the needed way.

**Comment by rvasin** [comment]
The issue was implemented in  #54737
